src/dotnet-grpc/Commands/CommandBase.cs

```diff
@@ -0,0 +1,358 @@
+        }
+
+        private void AddNugetPackage(string packageName, string packageVersion, bool privateAssets = false)
+        {
+            if (Project == null)
+            {
+                throw new InvalidOperationException("Internal error: Project not set.");
+            }
+
+            if (Console == null)
+            {
+                throw new InvalidOperationException("Internal error: Console not set.");
+            }
+
+            var packageReference = Project.GetItems("PackageReference").SingleOrDefault(i => i.UnevaluatedInclude == packageName);
+
+            if (packageReference == null)
+            {
+                Console.Out.WriteLine($"Adding required gRPC package reference: {packageName}.");
+
+                packageReference = Project.AddItem("PackageReference", packageName).Single();
+                packageReference.Xml.AddMetadata("Version", packageVersion, expressAsAttribute: true);
+
+                if (privateAssets)
+                {
+                    packageReference.Xml.AddMetadata("PrivateAssets", "All", expressAsAttribute: true);
+                }
+            }
+        }
+
+        public void AddProtobufReference(Services services, string additionalImportDirs, Access access, string file, string url)
+        {
+            if (Project == null)
+            {
+                throw new InvalidOperationException("Internal error: Project not set.");
+            }
+
+            if (Console == null)
+            {
+                throw new InvalidOperationException("Internal error: Console not set.");
+            }
+
+            if (!File.Exists(Path.IsPathRooted(file) ? file : Path.Join(Project.DirectoryPath, file)))
+            {
+                throw new CLIToolException($"The reference {file} does not exist.");
+            }
+
+            if (!Project.GetItems("Protobuf").Any(i => i.UnevaluatedInclude == file))
+            {
+                if (Path.GetExtension(file) != ".proto")
+                {
+                    Console.Out.WriteLine($"Warning: The reference {file} does not reference a .proto file. This may lead to compilation errors.");
+                }
+
+                var newItem = Project.AddItem("Protobuf", file).Single();
+
+                if (services != Services.Both)
+                {
+                    newItem.Xml.AddMetadata("GrpcServices", services.ToString(), expressAsAttribute: true);
+                }
+
+                if (access != Access.Public)
+                {
+                    newItem.Xml.AddMetadata("Access", access.ToString(), expressAsAttribute: true);
+                }
+
+                if (!string.IsNullOrEmpty(additionalImportDirs))
+                {
+                    newItem.Xml.AddMetadata("AdditionalImportDirs", additionalImportDirs, expressAsAttribute: true);
+                }
+
+                if (!string.IsNullOrEmpty(url))
+                {
+                    newItem.Xml.AddMetadata("SourceURL", url);
```

Comment:
Is SourceURL something we have come up with? Why SourceURL instead of SourceUrl?

Reply:
This name is pilfered from https://github.com/aspnet/AspNetCore/pull/10015/files. But yea I should change the casing.